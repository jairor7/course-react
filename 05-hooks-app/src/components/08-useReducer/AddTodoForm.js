import React from "react";
import PropTypes from "prop-types";
import { useForm } from "../../hooks/useForm";

const AddTodoForm = ({ handleAddTodo }) => {
  const [{ description }, handleInputChange, reset] =
    useForm({
      description: "",
    });

  const handleSubmit = (e) => {
    e.preventDefault();

    if (description.trim().length === 0) {
      return;
    }

    const newTodo = {
      id: new Date().getTime(),
      description,
      done: false,
    };
    handleAddTodo(newTodo);
    reset();
  };

  return (
    <>
      <h4>Agregar TODO</h4>
      <hr />
      <form onSubmit={handleSubmit}>
        <input
          type="text"
          placeholder="New TODO"
          name="description"
          autoComplete="off"
          className="form-control"
          value={description}
          onChange={handleInputChange}
        />
        <div className="d-grid">
          <button
            type="submit"
            className="btn btn-outline-primary btn-block mt-3"
          >
            Agregar
          </button>
        </div>
      </form>
    </>
  );
};

AddTodoForm.propTypes = {
  handleAddTodo: PropTypes.func.isRequired,
};

export default AddTodoForm;
