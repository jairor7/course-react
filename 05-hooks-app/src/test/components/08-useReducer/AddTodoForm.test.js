import AddTodoForm from "../../../components/08-useReducer/AddTodoForm";
import { shallow } from "enzyme";

describe("Pruebas al componente <AddTodoForm />", () => {
  const handleAddTodo = jest.fn();
  const wrapper = shallow(
    <AddTodoForm handleAddTodo={handleAddTodo} />
  );
  test("Debe mostrar correctamente el componente", () => {
    expect(wrapper).toMatchSnapshot();
  });

  test("No debe de llamar handleAddTodo", () => {
    //wrapper.find("form").simulate("submit");
    const formSubmit = wrapper
      .find("form")
      .prop("onSubmit"); //Otra forma de simular una acción como click o submit
    formSubmit({ preventDefault() {} });
    expect(handleAddTodo).toHaveBeenCalledTimes(0); //Veces que se llamo la función
  });

  test("Debe de llamar la función handleAddTodo con un argumento", () => {
    const valueDescription = "Aprender React";
    const inputChange = wrapper
      .find("input")
      .prop("onChange");
    inputChange({
      target: {
        value: valueDescription,
        name: "description",
      },
    });
    const formSubmit = wrapper
      .find("form")
      .prop("onSubmit");
    formSubmit({ preventDefault() {} });
    expect(handleAddTodo).toHaveBeenCalledWith({
      id: expect.any(Number),
      description: valueDescription,
      done: false,
    });
    expect(wrapper.find("input").prop("value")).toBe("");
  });
});
