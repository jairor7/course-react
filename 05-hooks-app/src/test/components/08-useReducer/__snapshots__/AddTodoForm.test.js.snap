// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas al componente <AddTodoForm /> Debe mostrar correctamente el componente 1`] = `
<Fragment>
  <h4>
    Agregar TODO
  </h4>
  <hr />
  <form
    onSubmit={[Function]}
  >
    <input
      autoComplete="off"
      className="form-control"
      name="description"
      onChange={[Function]}
      placeholder="New TODO"
      type="text"
      value=""
    />
    <div
      className="d-grid"
    >
      <button
        className="btn btn-outline-primary btn-block mt-3"
        type="submit"
      >
        Agregar
      </button>
    </div>
  </form>
</Fragment>
`;
